rfd/0152-automatic-database-users-mongodb.md

```diff
@@ -0,0 +1,202 @@
+implemented in Go. Multiple parallel database sessions will not race thanks to
+[semaphore
+locking](https://github.com/gravitational/teleport/blob/master/rfd/0113-automatic-database-users.md#locking).
+
+### Roles
+
+Unlike MySQL or PostgreSQL where roles are scoped to the entire database
+instance/cluster, a role in MongoDB is scoped to a specific database.
+
+For example, a custom role `myCustomRole` can be created on database `db1` with
+specific privileges on `db1`, and another role `myCustomRole` can be created on
+database `db2` with specific privileges on `db2`. They will be considered two
+different roles.
+
+Most built-in roles are available on all databases, while some built-in roles
+like `readAnyDatabase` can only be applied on the `admin` database.
+
+Therefore, when specifying database roles to assign for the user, it must be in
+the format of `<role-name>@<db-name>` to fully identify the role:
+
+```yaml
+kind: "role"
+version: "v6"
+metadata:
+  name: "example"
+spec:
+  options:
+    create_db_user_mode: keep
+  allow:
+    db_names:
+    - "db1"
+    - "db2"
+    - "db3"
+    db_roles: 
+    - "readAnyDatabase@admin"
+    - "readWrite@db2"
+    - "myCustomRole@db3"
+```
+
+Note that in the above example, even though the user is assigned role
+`readAnyDatabase@admin`, Teleport will block access to databases not in the
+`db_names`.
+
+### User accounts
+
+New users with name `CN=<teleport-username>` will be created on `$external`
+database to use X.509 authentication.
+
+All auto-provisioned users will be assigned role `teleport-auto-user@admin`.
+The `teleport-auto-user` is created on database `admin` with empty privileges.
+
+There is no built-in way to lock an user account in MongoDB, for deactivation
+purpose. As a workaround, the following `authenticationRestrictions` will be
+applied to the user account, in addition to stripping the roles:
+
+```json
+{
+  "updateUser": "CN=<teleport-username>",
+  "roles": [
+    { "role": "teleport-auto-user", "db": "admin" }
+  ],
+  "authenticationRestrictions": [
+    { "clientSource": ["0.0.0.0"] }
+  ]
+}
+```
+
+Limiting the `clientSource` effectively locks out the user from logging in.
+
+When re-activating the user, `clientSource` will be set to `["0.0.0.0/0"]`.
+
+### Finding active connections
+
+Command `currentOp` is used to find active connections for a specific user:
+```json
+{
+  "currentOp": true,
+  "$ownOps": false,
+  "$all": true,
+  "effectiveUsers": {
+    "$elemMatch": {
+      "user": "CN=<teleport-username>",
+      "db": "$external"
+    }
+  }
+}
+```
+
+## UX
+
+Database roles must be specified in format of `<role-name>@<db-name>`. See
+"Roles" section above for more details.
+
+## Security
+
+The admin user requires role `clusterMonitor@admin` and role
+`userAdminAnyDatabase@admin` to monitor connections and manage users.
+
+Auth restrictions `clientSource: ["0.0.0.0"]` is used to lock an user account
```

Comment:
This will only prevent future connections, correct?

Reply:
yes, similar to the other comment, this is to prevent someone from logging in as this database user (not through Teleport auto-user provisioning flow).